### PR TITLE
IGV compatible filtering of variants

### PR DESCRIPTION
**Pull request by mhkc**
This PR adds a new input field when filtering variants. The filter takes an IGV position string (<chr>:<start>-<end>) as input which is a feature frequently requested by the analysts at Clinical Genomics Lund. The intent is to enable copy-pasting in position information from genome browsers such as IGV directly into scout.

**How to test**
1. Load test case 643594 into scout.
2. Go to http://localhost:5000/cust000/643594/sv/variants
3. Try searches such as in Chromosome position field 9:1,8500,001-2,800,0001

**Expected outcome**:
1. The same type of filtering as when using the ordinary filters.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by dnil** [comment]
Nice; interoperability is a Good Thing, and I personally like free text fields much better than divided boxes. Have you considered adding a conditional "chr" to the regexps at the start of the chromosome name, for pasting from builds that default to names like that?

Screenshot 2020-11-10 at 17 06 27

**Comment by mhkc** [comment]
I can add the optional chr in the regex.

Btw I kept the separate chr, start and end fields since I dont know if some analyst prefer to have separate fields compared to one unified. Should we keep both input options for now?

**Comment by dnil** [comment]
I agree it is best to leave the old way in until current users have had a chance to try the new. Perhaps they will all realise it is easier and at least as safe just to type than click dropdowns, but I am not convinced they will see it that way. 😊

**Comment by mhkc** [comment]
Sounds good! What can I do to help with the merging of the PR?

**Comment by dnil** [comment]
We (@northwestwitch perhaps for the js, otw I can be a fallback) review to give you some feedback, but perhaps you wanted to extend the regexp to catch "chr". Also to forestall a review, have a look at the lint, test and code factor output. I have a feeling you may disagree style-wise with the js lint, but just note that if so.

**Comment by northwestwitch** [comment]
I can definitely give a look at this PR this week! :)

**Comment by mhkc** [comment]
Ill update the regex tomorrow :)

**Comment by northwestwitch** [comment]
I'm going to review this one!

File: scout/adapter/mongo/query.py
Comment (northwestwitch) [inline scout/adapter/mongo/query.py:183]: I've been testing the branch on a local instance and I see that every time a user inserts coordinates in the chrom_pos fields -> chromosome, start, and stop fields are also updated (not applying the eventual padding). If you apply the padding in the web page (at the same time you update these fields) then you don't have to touch this file at all, because you can perform the search using those values instead (only chromosome, start and stop). It would avoid a lot of additional code in a function that is already exploding (not your fault, it's already a monster function as it is)

File: scout/server/blueprints/variants/forms.py
Comment (northwestwitch) [inline scout/server/blueprints/variants/forms.py:108]: 👍

File: scout/server/blueprints/variants/static/form_scripts.js
Comment (northwestwitch) [inline scout/server/blueprints/variants/static/form_scripts.js:2]: Testing with coordinates I've noticed that the regex allows for searching for chromosomes ranging 23-99, which don't exist of course. For instance if you enter chromosome 99 the form is submitted and the chromosome on top is assigned to "9"

File: scout/server/blueprints/variants/static/form_scripts.js
Comment (northwestwitch) [inline scout/server/blueprints/variants/static/form_scripts.js:20]: Updating cytobands according to the values of chrom_pos is nice, but it's not really necessary for the subsequent search, so you can leave it out if you want.

File: scout/server/blueprints/variants/static/form_scripts.js
Comment (northwestwitch) [inline scout/server/blueprints/variants/static/form_scripts.js:84]: As I mentioned before, the pattern could be improved by checking that chromosome range is [1-22] and X,Y,MT.
Also the suggestion from @dnil (allowing "chr" as prefix) could be implemented

File: scout/server/blueprints/variants/static/form_scripts.js
Comment (northwestwitch) [inline scout/server/blueprints/variants/static/form_scripts.js:90]: I think it's safe to ignore codefactor here.

File: scout/server/blueprints/variants/static/form_scripts.js
Comment (northwestwitch) [inline scout/server/blueprints/variants/static/form_scripts.js:158]: Fix pattern, see above.

File: scout/server/blueprints/variants/static/form_scripts.js
Comment (northwestwitch) [inline scout/server/blueprints/variants/static/form_scripts.js:152]: Perhaps elaborate a bit more on this comment?

File: scout/server/blueprints/variants/static/form_scripts.js
Comment (northwestwitch) [inline scout/server/blueprints/variants/static/form_scripts.js:153]: What does this function do? I think it's very useful if:

- if the user enters values in the chrom_pos:
  - The chromosome select is updated
  - start is updated (with eventual padding)
  - end is updated (with eventual padding)

What I don't think is useful (at least for the subsequent filtering):
- Update chrom_pos  if a user modifies chrom and cytobands, also because chrom_pos doesn't update if you modify chromosome, start and stop

File: scout/server/blueprints/variants/templates/variants/utils.html
Comment (northwestwitch) [inline scout/server/blueprints/variants/templates/variants/utils.html:259]: This would look much nicer code-wise if you use a grid system like this: https://getbootstrap.com/docs/4.0/layout/grid/#variable-width-content

File: scout/server/blueprints/variants/templates/variants/utils.html
Comment (northwestwitch) [inline scout/server/blueprints/variants/templates/variants/utils.html:395]: Same here

File: scout/server/blueprints/variants/templates/variants/variants.html
Comment (northwestwitch) [inline scout/server/blueprints/variants/templates/variants/variants.html:179]: I've noticed that while the search field is present in the SNVs and SVs pages, it's missing in the search form of cancer variants and SV cancer variants. It would be very useful if you could add it there too

**Review by northwestwitch — CHANGES_REQUESTED** [review]
Nice feature and job! I think it would be very much appreciated by the users!

I have some suggestion for improvement, as you can see in the comments. 
I think the biggest change that I'd do is not touching the adapter/mongo/query function at all. Instead , given a chrom_pos string provided in the form, I would update properly the chrom, start and stop coordinates so that they will continue to be used in the search as it is now.

**Comment by northwestwitch** [comment]
Hi @mhkc let me know when you are ready for another review. Have a good day! :)

**Comment by mhkc** [comment]
@northwestwitch I have updated the things you mentioned so feel free to have a second review

**Comment by northwestwitch** [comment]
> @northwestwitch I have updated the things you mentioned so feel free to have a second review

Great, I'll review today!

**Comment by northwestwitch** [comment]
Hi @mhkc since this PR is changing the code and the variant filtering I think that it would be best if you could move it to a branch started from our repo, so we see if it passes the tests and the effect on the coverage. Could you do it? Otherwise I could do it for you!

**Comment by mhkc** [comment]
@northwestwitch if you could do it it would be great. I have a bunch of meetings today and would not have the time until tomorrow.

**Comment by northwestwitch** [comment]
I've moved this PR to COPY_filter-gene-by-c, closing this one!

**Pull request closed without merging**
